applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp

```diff
@@ -1122,6 +1317,67 @@ namespace Kratos
     //************************************************************************************
     //************************************************************************************
 
+    void BaseSolidElement::CalculateAndAddKm(
```

Comment:
I agree with both of you, not to duplicate the code and making the function efficient. Nevertheless, efficient implementation of this is somehow tricky. The good point is the fact that all matrices are in the cache so one the order of operation is only governed by vectorization.